File: android/app/src/main/java/com/ccoan/ny/ui/screens/meetings/MeetingsViewModel.kt
package com.ccoan.ny.ui.screens.meetings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ccoan.ny.core.network.ApiService
import com.ccoan.ny.data.models.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject

@HiltViewModel
class MeetingsViewModel @Inject constructor(
    private val apiService: ApiService
) : ViewModel() {

    private val _meetings = MutableStateFlow<List<Meeting>>(emptyList())
    val meetings: StateFlow<List<Meeting>> = _meetings.asStateFlow()

    private val _departments = MutableStateFlow<List<Department>>(emptyList())
    val departments: StateFlow<List<Department>> = _departments.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _selectedDepartmentId = MutableStateFlow<String?>(null)
    val selectedDepartmentId: StateFlow<String?> = _selectedDepartmentId.asStateFlow()

    // Calendar state
    private val _displayedMonth = MutableStateFlow(Date())
    val displayedMonth: StateFlow<Date> = _displayedMonth.asStateFlow()

    private val _selectedDate = MutableStateFlow<Date?>(null)
    val selectedDate: StateFlow<Date?> = _selectedDate.asStateFlow()

    private val _calendarDays = MutableStateFlow<List<Date>>(emptyList())
    val calendarDays: StateFlow<List<Date>> = _calendarDays.asStateFlow()

    private val calendar = Calendar.getInstance()

    init {
        updateCalendarDays()
    }

    // MARK: - Calendar Methods

    private fun updateCalendarDays() {
        val days = mutableListOf<Date>()
        val cal = Calendar.getInstance()
        cal.time = _displayedMonth.value

        // Go to first day of month
        cal.set(Calendar.DAY_OF_MONTH, 1)

        // Go back to Sunday of the first week
        val firstDayOfWeek = cal.get(Calendar.DAY_OF_WEEK)
        cal.add(Calendar.DAY_OF_MONTH, -(firstDayOfWeek - Calendar.SUNDAY))

        // Generate 42 days (6 weeks)
        repeat(42) {
            days.add(cal.time)
            cal.add(Calendar.DAY_OF_MONTH, 1)
        }

        _calendarDays.value = days
    }

    fun getMonthYearString(): String {
        val formatter = SimpleDateFormat("MMMM yyyy", Locale.US)
        return formatter.format(_displayedMonth.value)
    }

    fun goToPreviousMonth() {
        val cal = Calendar.getInstance()
        cal.time = _displayedMonth.value
        cal.add(Calendar.MONTH, -1)
        _displayedMonth.value = cal.time
        updateCalendarDays()
    }

    fun goToNextMonth() {
        val cal = Calendar.getInstance()
        cal.time = _displayedMonth.value
        cal.add(Calendar.MONTH, 1)
        _displayedMonth.value = cal.time
        updateCalendarDays()
    }

    fun goToCurrentMonth() {
        _displayedMonth.value = Date()
        _selectedDate.value = null
        updateCalendarDays()
    }

    fun selectDate(date: Date?) {
        _selectedDate.value = date
    }

    fun isCurrentMonth(date: Date): Boolean {
        val displayCal = Calendar.getInstance().apply { time = _displayedMonth.value }
        val dateCal = Calendar.getInstance().apply { time = date }
        return displayCal.get(Calendar.MONTH) == dateCal.get(Calendar.MONTH) &&
                displayCal.get(Calendar.YEAR) == dateCal.get(Calendar.YEAR)
    }

    fun isToday(date: Date): Boolean {
        val todayCal = Calendar.getInstance()
        val dateCal = Calendar.getInstance().apply { time = date }
        return todayCal.get(Calendar.YEAR) == dateCal.get(Calendar.YEAR) &&
                todayCal.get(Calendar.DAY_OF_YEAR) == dateCal.get(Calendar.DAY_OF_YEAR)
    }

    fun isSelected(date: Date): Boolean {
        val selected = _selectedDate.value ?: return false
        val selectedCal = Calendar.getInstance().apply { time = selected }
        val dateCal = Calendar.getInstance().apply { time = date }
        return selectedCal.get(Calendar.YEAR) == dateCal.get(Calendar.YEAR) &&
                selectedCal.get(Calendar.DAY_OF_YEAR) == dateCal.get(Calendar.DAY_OF_YEAR)
    }

    fun hasMeetings(date: Date): Boolean {
        val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        val dateString = dateFormatter.format(date)
        return getFilteredMeetings().any { it.date == dateString }
    }

    fun meetingsForDate(date: Date): List<Meeting> {
        val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        val dateString = dateFormatter.format(date)
        return getFilteredMeetings()
            .filter { it.date == dateString }
            .sortedBy { it.startTime }
    }

    fun formatSelectedDate(date: Date): String {
        val formatter = SimpleDateFormat("EEEE, MMMM d, yyyy", Locale.US)
        return formatter.format(date)
    }

    // MARK: - Filtering

    private fun getFilteredMeetings(): List<Meeting> {
        val deptId = _selectedDepartmentId.value
        return if (deptId != null) {
            _meetings.value.filter { it.departmentId == deptId }
        } else {
            _meetings.value
        }
    }

    fun getUpcomingMeetings(): List<Meeting> {
        return getFilteredMeetings()
            .filter { it.isUpcoming }
            .sortedBy { it.date }
    }

    fun getPastMeetings(): List<Meeting> {
        return getFilteredMeetings()
            .filter { !it.isUpcoming }
            .sortedByDescending { it.date }
    }

    fun loadData() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                val meetingsResponse = apiService.getMeetings()
                if (meetingsResponse.isSuccessful) {
                    _meetings.value = meetingsResponse.body()?.data ?: emptyList()
                }

                val deptResponse = apiService.getDepartments()
                if (deptResponse.isSuccessful) {
                    _departments.value = deptResponse.body()?.data ?: emptyList()
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }

            _isLoading.value = false
        }
    }

    fun setDepartmentFilter(deptId: String?) {
        _selectedDepartmentId.value = deptId
    }

    fun createMeeting(request: CreateMeetingRequest, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val response = apiService.createMeeting(request)
                if (response.isSuccessful && response.body()?.success == true) {
                    loadData()
                    onSuccess()
                } else {
                    _errorMessage.value = response.body()?.error ?: "Failed to create meeting"
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
            _isLoading.value = false
        }
    }

    fun updateMeeting(request: UpdateMeetingRequest, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val response = apiService.updateMeeting(request)
                if (response.isSuccessful) {
                    loadData()
                    onSuccess()
                } else {
                    _errorMessage.value = "Failed to update meeting"
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
            _isLoading.value = false
        }
    }

    fun deleteMeeting(id: String) {
        viewModelScope.launch {
            try {
                val response = apiService.deleteMeeting(id)
                if (response.isSuccessful) {
                    _meetings.value = _meetings.value.filter { it.id != id }
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }
}

File: ios/CCOAN/Features/Meetings/ViewModels/MeetingsViewModel.swift
import Foundation

@MainActor
class MeetingsViewModel: ObservableObject {
    @Published var meetings: [Meeting] = []
    @Published var departments: [Department] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var selectedDepartmentId: String?

    // Calendar state - weekly view
    @Published var currentWeekStart: Date = Date()
    @Published var weekDays: [Date] = []
    @Published var selectedDate: Date?

    private let apiClient = APIClient.shared
    private let calendar = Calendar.current

    init() {
        currentWeekStart = getWeekStart(from: Date())
        updateWeekDays()
    }

    // MARK: - Calendar Methods

    private func getWeekStart(from date: Date) -> Date {
        var cal = calendar
        cal.firstWeekday = 1 // Sunday
        let components = cal.dateComponents([.yearForWeekOfYear, .weekOfYear], from: date)
        return cal.date(from: components) ?? date
    }

    func updateWeekDays() {
        var days: [Date] = []
        for i in 0..<7 {
            if let date = calendar.date(byAdding: .day, value: i, to: currentWeekStart) {
                days.append(date)
            }
        }
        weekDays = days
    }

    var weekRangeString: String {
        guard let firstDay = weekDays.first, let lastDay = weekDays.last else { return "" }
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        let startStr = formatter.string(from: firstDay)
        formatter.dateFormat = "MMM d, yyyy"
        let endStr = formatter.string(from: lastDay)
        return "\(startStr) - \(endStr)"
    }

    func goToPreviousWeek() {
        if let newStart = calendar.date(byAdding: .day, value: -7, to: currentWeekStart) {
            currentWeekStart = newStart
            updateWeekDays()
        }
    }

    func goToNextWeek() {
        if let newStart = calendar.date(byAdding: .day, value: 7, to: currentWeekStart) {
            currentWeekStart = newStart
            updateWeekDays()
        }
    }

    func goToCurrentWeek() {
        currentWeekStart = getWeekStart(from: Date())
        updateWeekDays()
        selectedDate = nil
    }

    func selectDate(_ date: Date?) {
        selectedDate = date
    }

    func isToday(_ date: Date) -> Bool {
        calendar.isDateInToday(date)
    }

    func isSelected(_ date: Date) -> Bool {
        guard let selected = selectedDate else { return false }
        return calendar.isDate(date, inSameDayAs: selected)
    }

    func meetingsForDate(_ date: Date) -> [Meeting] {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let dateString = dateFormatter.string(from: date)
        return filteredMeetings.filter { $0.date == dateString }.sorted { $0.startTime < $1.startTime }
    }

    func formatDayHeader(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEE"
        return formatter.string(from: date)
    }

    func formatDayNumber(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "d"
        return formatter.string(from: date)
    }

    func formatFullDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE, MMMM d, yyyy"
        return formatter.string(from: date)
    }

    // MARK: - Filtering

    private var filteredMeetings: [Meeting] {
        meetings.filter { meeting in
            if let deptId = selectedDepartmentId {
                return meeting.departmentId == deptId
            }
            return true
        }
    }

    var upcomingMeetings: [Meeting] {
        filteredMeetings
            .filter { $0.isUpcoming }
            .sorted { $0.date < $1.date }
    }

    var pastMeetings: [Meeting] {
        filteredMeetings
            .filter { !$0.isUpcoming }
            .sorted { $0.date > $1.date }
    }

    func loadData() async {
        isLoading = true
        errorMessage = nil

        do {
            async let meetingsResponse: APIResponse<[Meeting]> = apiClient.get(endpoint: .meetings)
            async let departmentsResponse: APIResponse<[Department]> = apiClient.get(endpoint: .departments)

            let (meetRes, deptRes) = try await (meetingsResponse, departmentsResponse)

            if let meetingsData = meetRes.data {
                self.meetings = meetingsData
            }
            if let departmentsData = deptRes.data {
                self.departments = departmentsData
            }
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func createMeeting(_ request: CreateMeetingRequest) async -> Bool {
        do {
            let response: APIResponse<Meeting> = try await apiClient.post(
                endpoint: .meetings,
                body: request
            )
            if response.success {
                await loadData()
                return true
            } else {
                errorMessage = response.error ?? "Failed to create meeting"
            }
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }

    func updateMeeting(_ request: UpdateMeetingRequest) async -> Bool {
        do {
            let response: APIResponse<EmptyResponse> = try await apiClient.put(
                endpoint: .meetings,
                body: request
            )
            if response.success {
                await loadData()
                return true
            } else {
                errorMessage = response.error ?? "Failed to update meeting"
            }
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }

    func deleteMeeting(id: String) async -> Bool {
        do {
            let response: APIResponse<EmptyResponse> = try await apiClient.delete(
                endpoint: .meetings,
                queryParams: ["id": id]
            )
            if response.success {
                meetings.removeAll { $0.id == id }
                return true
            }
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }
}
